File: Easy_Leetcode_Problems/Easy_Leetcode_Problems/Maximum_product_of_two_digits_3536.java
// Logic:
//
// 1. Convert the number into a String.
//
// 2. Traverse every pair of digits.
//
// 3. Convert characters into integers.
//
// 4. Find the product of every pair.
//
// 5. Store the maximum product.
//
// 6. Return the maximum product.

// Algorithm:
//
// 1. Convert n to String.
//
// 2. Initialize maxProduct = 0.
//
// 3. Use two nested loops.
//
// 4. Convert each character to digit.
//
// 5. Calculate:
//      product = digit1 * digit2
//
// 6. Update maxProduct.
//
// 7. Return maxProduct.

class Solution {
    public int maxProduct(int n) {

        String num = String.valueOf(n);

        int maxProduct = 0;

        for (int i = 0; i < num.length(); i++) {
            for (int j = i + 1; j < num.length(); j++) {

                int digit1 = num.charAt(i) - '0';
                int digit2 = num.charAt(j) - '0';

                maxProduct = Math.max(maxProduct, digit1 * digit2);
            }
        }

        return maxProduct;
    }
}


// ⭐ Optimized Java Solution (Without Converting to String)

class Solution {
    public int maxProduct(int n) {

        int first = 0;
        int second = 0;

        while (n > 0) {
            int digit = n % 10;

            if (digit >= first) {
                second = first;
                first = digit;
            } else if (digit > second) {
                second = digit;
            }

            n /= 10;
        }

        return first * second;
    }
}

// Interview Explanation:
//
// 1. I converted the number into a String.
//
// 2. I used two nested loops to
//    compare every pair of digits.
//
// 3. I converted each character
//    into an integer.
//
// 4. I calculated the product of
//    every pair and kept the
//    maximum product.
//
// 5. Finally, I returned the
//    maximum product.
//
// Time Complexity: O(d²)
// Space Complexity: O(d)
//
// where d = number of digits in n.

File: Easy_Leetcode_Problems/Easy_Leetcode_Problems/Maximum_product_of_two_digits_3536.py
# Logic:
#
# 1. Convert the number into a string.
#
# 2. Store every digit in a list.
#
# 3. Check every pair of digits.
#
# 4. Calculate their product.
#
# 5. Keep updating the maximum product.
#
# 6. Return the maximum product.

# Algorithm:
#
# 1. Convert n into a string.
#
# 2. Convert each character into an integer.
#
# 3. Store all digits in a list.
#
# 4. Initialize maxProduct = 0.
#
# 5. Use two loops to check every pair.
#
# 6. Compute:
#       product = digits[i] * digits[j]
#
# 7. Update maxProduct if needed.
#
# 8. Return maxProduct.

class Solution:
    def maxProduct(self, n: int) -> int:

        digits = [int(d) for d in str(n)]

        maximum = 0

        for i in range(len(digits)):
            for j in range(i + 1, len(digits)):
                maximum = max(maximum, digits[i] * digits[j])

        return maximum

# Interview Explanation:
#
# 1. I converted the number into a list of digits.
#
# 2. I used two nested loops to check every
#    possible pair of digits.
#
# 3. For each pair, I calculated the product
#    and kept track of the maximum value.
#
# 4. Finally, I returned the maximum product.
#
# Time Complexity: O(d²)
# Space Complexity: O(d)
#
# where d is the number of digits in n.